### ISSUE TITLE
Semaphore timeout upon dumping USB drive

### ISSUE THREAD

**Issue by Feathered-Serpent** [saramibreak/DiscImageCreator#124]
**Version**
32 bit, AnsiBuild, 20220331T120308

**Describe the bug**
I played around some with dumping a USB drive. My first one (1GB) was alright, now my second brings me this error:
```
AppVersion
        32 bit, AnsiBuild, 20220331T120308
CurrentDirectory
        C:\Users\Ezatoka\Desktop\UFD_3.0 Silicon-Power USB Device
WorkingPath
         Argument: UFD_3.0 Silicon-Power USB Device.bin
         FullPath: C:\Users\Ezatoka\Desktop\UFD_3.0 Silicon-Power USB Device\UFD_3.0 Silicon-Power USB Device.bin
            Drive: C:
        Directory: \Users\Ezatoka\Desktop\UFD_3.0 Silicon-Power USB Device\
         Filename: UFD_3.0 Silicon-Power USB Device
        Extension: .bin
StartTime: 2022-04-04T12:50:10+0200
DiskSize: 15550382592 bytes, BytesPerSector: 512, BlockSize: 30371841
[F:Read10][L:114] GetLastError: 121, The semaphore timeout period has expired.
```

**Log file**
[UFD_3.0 Silicon-Power USB Device.zip](https://github.com/saramibreak/DiscImageCreator/files/8409014/UFD_3.0.Silicon-Power.USB.Device.zip)

**Comment by Feathered-Serpent** [comment]
Uhm how do I even dump a USB drive correctly? I used "DiscImageCreator.exe disk F test.bin". I tried with yet another USB stick, 32GB in size which at this point had nearly 10GB data on it. From the 31608274944 Bytes in total it loaded around 6 million in the console (stating it loaded a block), then stated it was done. I have a 32GB file, but it's all empty. If creating a zip file from it, the zip has around 64 MB in size.

Logs:
[JetFlash Transcend 32GB USB Device.zip](https://github.com/saramibreak/DiscImageCreator/files/8409743/JetFlash.Transcend.32GB.USB.Device.zip)

**Comment by saramibreak** [comment]
> I have a 32GB file, but it's all empty.

I also dumped 32 and 16 GB usb drive, it's no problem and it's can mount by IsoBuster.

**Comment by Feathered-Serpent** [comment]
Today I tried to run DiscImageCreator disk E test.bin again and strangely this time it worked. I have no clue, why. All that changed was a power down/up from yesterday to today. I also now recognized, that BlockSize: 61734912 times BytesPerSector: 512 equals DiskSize: 31608274944 bytes. It's just a bit confusing (for me at least).

Btw wouldn't it be better to let the hash be created during dumping? Seems possible with DVD/BD according to changelog ("- changed: Hash image while dumping (DVD, BD)") so it might be possible there as well. 

**But** my problem from the first post stays: even with the other USB stick working now  to be dumped, the other one still shows the semaphore window.

I open a command line, plug the USB stick into the USB port and run the dump command. I'm not doing anything prior that to the USB stick. The other log files (_disc and _drive) show that DIC recognized the stick, but somehow it can't start dumping it.

**Comment by saramibreak** [comment]
> Btw wouldn't it be better to let the hash be created during dumping? Seems possible with DVD/BD according to changelog ("- changed: Hash image while dumping (DVD, BD)") so it might be possible there as well.

Yes, I'll change to it.

> But my problem from the first post stays

Try also other software (e.g. IsoBuster), plz.

**Comment by Feathered-Serpent** [comment]
Not sure if these informations can be of any use.

USBDeview info from a working 1GB stick:
![working](https://user-images.githubusercontent.com/72162369/161796958-59ea713a-baad-43e8-ad38-064cfa71d2c2.png)

USBDeview info from the non-working 16GB stick:
![not working](https://user-images.githubusercontent.com/72162369/161796995-5cb200e6-b9ed-43c6-80cf-34e3ab381f34.png)

**Comment by Feathered-Serpent** [comment]
Two more sticks, one is working, the other one is giving me the Semaphore error.

Interestingly the stick not working has the Capability UniqueD just like the non-working 16 GB stick. But of course I have no clue if that says something.

Working: 
[8 GB Intenso Alu Line USB Device.zip](https://github.com/saramibreak/DiscImageCreator/files/8547646/8.GB.Intenso.Alu.Line.USB.Device.zip)
Not working: 
[128 GB Intenso Speed Line USB Device.zip](https://github.com/saramibreak/DiscImageCreator/files/8547647/128.GB.Intenso.Speed.Line.USB.Device.zip)

**Comment by Feathered-Serpent** [comment]
with ProcessMonitor I filtered to Operation DeviceIoControl and see, that there's an IO timeout
![ProcessMonitor](https://user-images.githubusercontent.com/72162369/164945087-d2513752-0e6c-4564-911c-28e24972e206.png)
strangely enough the operation IOCTL_SCSI_PASS_THROUGH_DIRECT was used before without problems hmm...

**Comment by saramibreak** [comment]
[DiscImageCreator_test.zip](https://github.com/saramibreak/DiscImageCreator/files/8548960/DiscImageCreator_test.zip)
- fixed: transfer length

**Comment by Feathered-Serpent** [comment]
There come a few errors upon start of dump:
```
C:\Users\Ezatoka\Desktop\Release_ANSI>DiscImageCreator.exe disk F nonworking16gb.bin
AppVersion
        32 bit, AnsiBuild, 20220424T115325
CurrentDirectory
        C:\Users\Ezatoka\Desktop\Release_ANSI
WorkingPath
         Argument: nonworking16gb.bin
         FullPath: C:\Users\Ezatoka\Desktop\Release_ANSI\nonworking16gb.bin
            Drive: C:
        Directory: \Users\Ezatoka\Desktop\Release_ANSI\
         Filename: nonworking16gb
        Extension: .bin
StartTime: 2022-04-24T10:49:09+0200
DiskSize: 15550382592 bytes, BytesPerSector: 512, BlockSize: 30371841
[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

Creating bin (Block) 30371841/30371841
EndTime: 2022-04-24T10:54:43+0200
```

```
C:\Users\Ezatoka\Desktop\Release_ANSI>DiscImageCreator.exe disk F nonworking128gb.bin
AppVersion
        32 bit, AnsiBuild, 20220424T115325
CurrentDirectory
        C:\Users\Ezatoka\Desktop\Release_ANSI
WorkingPath
         Argument: nonworking128gb.bin
         FullPath: C:\Users\Ezatoka\Desktop\Release_ANSI\nonworking128gb.bin
            Drive: C:
        Directory: \Users\Ezatoka\Desktop\Release_ANSI\
         Filename: nonworking128gb
        Extension: .bin
StartTime: 2022-04-24T11:14:53+0200
LBA[000000, 0000000]: [F:ModeSense10][L:719]
        Opcode: 0x5a
        ScsiStatus: 0x02 = CHECK_CONDITION
        SenseData Key-Asc-Ascq: 05-20-00 = ILLEGAL_REQUEST - INVALID COMMAND OPERATION CODE
DiskSize: 128094044160 bytes, BytesPerSector: 512, BlockSize: 250183680
[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

Creating bin (Block) 250183680/250183680
EndTime: 2022-04-24T11:52:56+0200
```

but both non-working sticks dump now. Strangely enough, the error about Wrong Parameter now also comes on the 1GB stick where it didn't came before. (it still dumps though)

```
AppVersion
        32 bit, AnsiBuild, 20220424T115325
CurrentDirectory
        C:\Users\Ezatoka\Desktop\Release_ANSI
WorkingPath
         Argument: working1gb.bin
         FullPath: C:\Users\Ezatoka\Desktop\Release_ANSI\working1gb.bin
            Drive: C:
        Directory: \Users\Ezatoka\Desktop\Release_ANSI\
         Filename: working1gb
        Extension: .bin
StartTime: 2022-04-24T11:55:46+0200
LBA[000000, 0000000]: [F:ModeSense10][L:719]
        Opcode: 0x5a
        ScsiStatus: 0x02 = CHECK_CONDITION
        SenseData Key-Asc-Ascq: 05-24-00 = ILLEGAL_REQUEST - INVALID FIELD IN CDB
DiskSize: 1007681536 bytes, BytesPerSector: 512, BlockSize: 1968128
[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

[F:OutputFsFATLDirEntry][L:840] GetLastError: 87, Falscher Parameter.

Creating bin (Block) 1968128/1968128
EndTime: 2022-04-24T11:56:45+0200
```

**Comment by saramibreak** [comment]
[DiscImageCreator_test.zip](https://github.com/saramibreak/DiscImageCreator/files/8553528/DiscImageCreator_test.zip)
- fixed: the bug when added [#125](https://github.com/saramibreak/DiscImageCreator/commit/eab7480b82a791b1d6158b0a6f4a11787cc4bf9c)

**Comment by Feathered-Serpent** [comment]
Working fine now with the so far non-working sticks!
